src/librbd/deep_copy/ObjectCopyRequest.cc

```diff
@@ -0,0 +1,660 @@
+    // possible that exclusive lock was lost in background
+    lderr(m_cct) << "object map is not initialized" << dendl;
+
+    m_dst_image_ctx->snap_lock.put_read();
+    m_dst_image_ctx->owner_lock.put_read();
+    finish(-EINVAL);
+    return;
+  }
+
+  auto &dst_object_state = *m_dst_object_state.begin();
+  auto it = m_snap_map.find(dst_object_state.first);
+  assert(it != m_snap_map.end());
+  auto dst_snap_id = it->second.front();
+  auto object_state = dst_object_state.second;
+  m_dst_object_state.erase(m_dst_object_state.begin());
+
+  ldout(m_cct, 20) << "dst_snap_id=" << dst_snap_id << ", object_state="
+                   << static_cast<uint32_t>(object_state) << dendl;
+
+  auto finish_op_ctx = start_lock_op(m_dst_image_ctx->owner_lock);
+  if (finish_op_ctx == nullptr) {
+    lderr(m_cct) << "lost exclusive lock" << dendl;
+    m_dst_image_ctx->snap_lock.put_read();
+    m_dst_image_ctx->owner_lock.put_read();
+    finish(-EROFS);
+    return;
+  }
+
+  auto ctx = new FunctionContext([this, finish_op_ctx](int r) {
+      handle_update_object_map(r);
+      finish_op_ctx->complete(0);
+    });
+
+  RWLock::WLocker object_map_locker(m_dst_image_ctx->object_map_lock);
+  bool sent = m_dst_image_ctx->object_map->template aio_update<
+    Context, &Context::complete>(dst_snap_id, m_dst_object_number, object_state,
+                                 {}, {}, ctx);
+  m_dst_image_ctx->snap_lock.put_read();
+  m_dst_image_ctx->owner_lock.put_read();
+  if (!sent) {
+    assert(dst_snap_id == CEPH_NOSNAP);
+    handle_update_object_map(0);
+  }
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::handle_update_object_map(int r) {
+  ldout(m_cct, 20) << "r=" << r << dendl;
+
+  assert(r == 0);
+  if (!m_dst_object_state.empty()) {
+    send_update_object_map();
+    return;
+  }
+  finish(0);
+}
+
+template <typename I>
+Context *ObjectCopyRequest<I>::start_lock_op(RWLock &owner_lock) {
+  assert(m_dst_image_ctx->owner_lock.is_locked());
+  if (m_dst_image_ctx->exclusive_lock == nullptr) {
+    return new FunctionContext([](int r) {});
+  }
+  return m_dst_image_ctx->exclusive_lock->start_op();
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::compute_src_object_extents() {
+  std::vector<std::pair<uint64_t, uint64_t>> image_extents;
+  Striper::extent_to_file(m_cct, &m_dst_image_ctx->layout, m_dst_object_number,
+                          0, m_dst_image_ctx->layout.object_size, image_extents);
+
+  size_t total = 0;
+  for (auto &e : image_extents) {
+    std::map<object_t, std::vector<ObjectExtent>> src_object_extents;
+    Striper::file_to_extents(m_cct, m_src_image_ctx->format_string,
+                             &m_src_image_ctx->layout, e.first, e.second, 0,
+                             src_object_extents);
+    for (auto &p : src_object_extents) {
+      for (auto &s : p.second) {
+        m_src_objects.insert(s.objectno);
+        m_src_object_extents.push_back({s.objectno, s.offset, s.length});
```

Comment:
@dillaman I think I have found an issue here.
I need `m_src_object_extents` list to be ordered by a destination object offset (so I could sequentially write extents to the destination object; alternatively I would need to track src_offset->dst_offset). But I did not realize `Striper::file_to_extents` returned a map ordered by object and the order is lost. So I need to sort extents returned in `src_object_extents`.

Right now I don't see a good way. It looks like using `Striper::extent_to_file` + `Striper::file_to_extents` like above I can't map `dst_object_offset` -> `src_object, src_object_offset`.

Reply:
Ah, it looks like I just need to use a different version of `Striper::file_to_extents`.

Reply:
No, the version that returns a list, uses `assimilate_extents` which does the same that my code above, so it returns the list in wrong order (

It looks I would need to extend `Striper` to provide `file_to_extents` version I need?

Reply:
@dillaman I fixed this by introducing `src_to_dst_object_offset()` (see the fixup commit 48029c2). Do you think it could be a solution or do we need something more efficient (e.g. update `Striper` to return results we need)?

After this change `fsx` still fails but in different place. Need to track this further.

Reply:
@dillaman Another problem with `src_to_dst_object_offset()` found: I need to split src object extents, if the destination stripe_unit is smaller than the source stripe_unit. The fixup commit 9024db9 contains the fix for this issue too. Now it is much better: I have not been able to make fsx fail so far!

I will squash the fixup after your review.